### PR TITLE
CRM-20359 Extended the use of contribution tokens to include custom fields using API

### PR DESCRIPTION
**Pull request by Edzelopez**
* [CRM-20359: Support contribution.custom_nn tokens for Thank You letters - print or email](https://issues.civicrm.org/jira/browse/CRM-20359)

### CODE REVIEW

**Comment by Edzelopez** [comment]
@monishdeb could you please QA?

**Comment by monishdeb** [comment]
@Edzelopez I have reviewed and updated your PR with additional changes. Have a look

**Comment by Edzelopez** [comment]
Changes look good, is this ready to merge?

**Comment by monishdeb** [comment]
@colemanw @eileenmcnaughton the patch here is to support contribution custom field token using Contribution.Get API instead of  ```CRM_Utils_Token::getContributionTokenDetails(..)```. Can you please merge it if the patch looks good?

File: CRM/Contribute/Form/Task/PDFLetterCommon.php [hunk 1]
```diff
@@ -247,13 +247,13 @@ public static function buildContributionArray($groupBy, $contributionIDs, $retur
...
-        NULL,
-        $messageToken,
-        $task
-      );
-      $contribution = $contributions[$contributionId] = $contribution[$contributionId];
+
+      // retrieve contribution tokens listed in $messageToken using Contribution.Get API
+      $contribution = civicrm_api3('Contribution', 'getsingle', array(
+        'id' => $contributionId,
+        'return' => CRM_Utils_Array::value('contribution', $messageToken, array()),
```
Comment (colemanw) [inline CRM/Contribute/Form/Task/PDFLetterCommon.php]: If `$messageToken['contribution']` is empty then this api call will return nothing. Maybe we should wrap the whole thing in a conditional and skip ti altogether in that case?

File: CRM/Contribute/Form/Task/PDFLetterCommon.php [hunk 2]
```diff
@@ -247,13 +247,15 @@ public static function buildContributionArray($groupBy, $contributionIDs, $retur
...
-        $returnProperties,
-        NULL,
-        $messageToken,
-        $task
-      );
-      $contribution = $contributions[$contributionId] = $contribution[$contributionId];
+
+      // retrieve contribution tokens listed in $messageToken using Contribution.Get API
+      $contributions[$contributionId] = array();
+      if (!empty($messageToken['contribution'])) {
```
Comment (monishdeb) [inline CRM/Contribute/Form/Task/PDFLetterCommon.php]: @colemanw thanks for your suggestion. Does it look good now?

Comment (colemanw) [inline CRM/Contribute/Form/Task/PDFLetterCommon.php]: Yes it looks good @monishdeb - is there a test that covers this?

File: tests/phpunit/CRM/Contribute/Form/Task/PDFLetterCommonTest.php [hunk 3]
```diff
@@ -62,19 +62,55 @@ public function tearDown() {
    */
   public function testBuildContributionArray() {
```
Comment (monishdeb) [inline tests/phpunit/CRM/Contribute/Form/Task/PDFLetterCommonTest.php]: @colemanw I have extended the unit test to assert custom field token.

File: CRM/Contribute/Form/Task/PDFLetterCommon.php [hunk 4]
```diff
@@ -247,13 +247,18 @@ public static function buildContributionArray($groupBy, $contributionIDs, $retur
...
-      $contribution = CRM_Utils_Token::getContributionTokenDetails(array('contribution_id' => $contributionId),
-        $returnProperties,
-        NULL,
-        $messageToken,
-        $task
-      );
-      $contribution = $contributions[$contributionId] = $contribution[$contributionId];
+
+      // basic return attributes needed, see below for there usage
+      $returnValues = array('contact_id', 'total_amount');
```
Comment (monishdeb) [inline CRM/Contribute/Form/Task/PDFLetterCommon.php]: There need to be two return attributes as required later in the function.

**Pull request merged**